keps/sig-auth/5681-conditional-authorization/README.md

```diff
@@ -0,0 +1,1425 @@
+   AdmissionReview webhooks, so the authorizer can know which conditions to
+   apply to which request.  
+   1. This approach does not have the “Two full evaluations needed” and
+      “Non-atomicity” problems of the first alternative approach, as only the
+      conditions need to be atomically evaluated against the resource data.
+      However, this approach is subject to the “Tight coupling” and “Not
+      observable through (Self)SubjectAccessReview” problems. In addition, the
+      following problems arise:  
+   2. **A stateful authorizer is complex and hard to scale:** The authorizer
+      must be way more complex, as it needs to keep a lookup table of request ID
+      to condition set internally. If the authorizer needs to be horizontally
+      scaled, the load balancer in front of the horizontally scaled authorizers
+      would somehow need to know which authorizer replica has what requests’
+      conditions stored in memory.
+      1. With this proposal, the authorizer is allowed to be stateless and thus
+         simpler. Therefore, also the horizontal scaling can be done in a
+         straightforward manner, from this perspective.
+   3. **Unclear caching semantics**: The authorizer would need to cache the
+      conditions in memory for at least as long as SubjectAccessReview requests
+      can be cached, for the above atomicity invariant to hold. However, the
+      authorizer does not (generally) know the API server configuration, and
+      thus does not know how long to cache the conditions, or if at all.
+
+### Condition and ConditionSet data model
+
+The data model is shown in Go pseudocode. These types and interfaces could be
+put in
+[k8s.io/apiserver/pkg/authorization/authorizer](http://k8s.io/apiserver/pkg/authorization/authorizer),
+or in a sub-package. They are abstract, non-serializable types, which concrete
+implementations implement or convert to/from.
+
+```go
+package authorizer
+
+// ConditionEffect specifies how a condition evaluating to true should be handled.
+type ConditionEffect string
+
+const (
+    // If any Deny condition evaluates to true, the ConditionSet 
+    // necessarily evaluates to Deny. No further authorizers 
+    // are consulted.
+    EffectDeny ConditionEffect = "Deny"
+    // If a NoOpinion condition evaluates to true, the given 
+    // authorizer’s ConditionSet cannot evaluate to Allow anymore, but 
+    // necessarily Deny or NoOpinion, depending on whether there are any
+    // true EffectDeny conditions. 
+    // However, later authorizers in the chain can still Allow or Deny.
+    // It is effectively a softer deny that just overrides the 
+    // authorizer's own allow policies. It can be used if an authorizer  
+    // does not consider itself fully authoritative for a given request.  
+    // TODO: Talk about error handling; what happens if any of these 
+    // conditions fail to evaluate.
+    EffectNoOpinion ConditionEffect = "NoOpinion"
```

Comment:
As a reader, it might be helpful to have some examples of when an authorizer presenting a condition to be evaluated may not consider itself authoritative.

Probably naively, my understanding of `NoOpinion` in the existing authorizer sense aligns with "I don't have an enforceable policy defined for this request, and thus have no opinion on the authorization of this request".

If an authorizer is returning a condition to be evaluated for a policy, I would assume that the authorizer inherently has an opinion on whether or not the request should be authorized based on the result of that condition evaluation?